linera-core/src/local_node.rs

```diff
@@ -393,36 +393,65 @@ where
         name: ValidatorName,
         mut node: A,
         chain_id: ChainId,
-        start: BlockHeight,
+        mut start: BlockHeight,
         stop: BlockHeight,
     ) -> Result<(), LocalNodeError>
     where
         A: LocalValidatorNode + Clone + 'static,
     {
-        let limit = u64::from(stop)
-            .checked_sub(u64::from(start))
-            .ok_or(ArithmeticError::Overflow)?;
+        while start < stop {
+            // TODO(#2045): Analyze network errors instead of guessing the batch size.
+            let limit = u64::from(stop)
+                .checked_sub(u64::from(start))
+                .ok_or(ArithmeticError::Overflow)?
+                .min(1000);
+            let Some(certificates) = self
+                .try_query_certificates_from(name, &mut node, chain_id, start, limit)
+                .await?
+            else {
+                break;
+            };
+            let Some(info) = self
+                .try_process_certificates(name, &mut node, chain_id, certificates)
+                .await
+            else {
+                break;
+            };
+            assert!(info.next_block_height > start);
```

Comment:
We shouldn't panic just because one validator sent us e.g. an empty chain info. I'd rather return an error here.

Reply:
I believe the loop breaks first in this case

Reply:
oh shoot, I see